File: 03_use_cases/13_macos_cua/python/metacua/bash_tool.py
"""The `bash` computer-use tool: run a bounded shell command."""

import signal
import subprocess
from typing import Tuple

from .errors import CLIError

_BASH_TOOL_MAX_OUTPUT_BYTES = 128 * 1024


def _truncate(data: bytes) -> Tuple[str, int]:
    """Cap captured output to the byte budget and report how much was dropped."""
    dropped = max(0, len(data) - _BASH_TOOL_MAX_OUTPUT_BYTES)
    kept = data[:_BASH_TOOL_MAX_OUTPUT_BYTES]
    text = kept.decode("utf-8", errors="replace")
    return text, dropped


def run_bash_tool(command: str, timeout_ms: int) -> str:
    trimmed = command.strip()
    if not trimmed:
        raise CLIError("bash command must not be empty")

    process = subprocess.Popen(
        ["/bin/bash", "-lc", command],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    timed_out = False
    try:
        out_data, err_data = process.communicate(timeout=timeout_ms / 1000.0)
    except subprocess.TimeoutExpired:
        timed_out = True
        process.terminate()
        try:
            out_data, err_data = process.communicate(timeout=2.0)
        except subprocess.TimeoutExpired:
            process.kill()
            try:
                out_data, err_data = process.communicate(timeout=2.0)
            except subprocess.TimeoutExpired:
                out_data, err_data = b"", b""

    exit_code = process.returncode if process.returncode is not None else -signal.SIGKILL

    result = _format_bash_tool_result(
        command=trimmed,
        timeout_ms=timeout_ms,
        timed_out=timed_out,
        exit_code=exit_code,
        stdout=_truncate(out_data or b""),
        stderr=_truncate(err_data or b""),
    )

    if timed_out or exit_code != 0:
        raise CLIError(result)
    return result


def _format_bash_tool_result(
    command: str,
    timeout_ms: int,
    timed_out: bool,
    exit_code: int,
    stdout: Tuple[str, int],
    stderr: Tuple[str, int],
) -> str:
    parts = [f"command: {command}"]
    if timed_out:
        parts.append(f"status: timed out after {timeout_ms}ms")
    else:
        parts.append(f"status: exit {exit_code}")
    parts.append(_format_stream("stdout", stdout))
    parts.append(_format_stream("stderr", stderr))
    return "\n".join(parts)


def _format_stream(name: str, stream: Tuple[str, int]) -> str:
    text, dropped = stream
    rendered = text.strip("\n")
    rendered = rendered if rendered else "<empty>"
    if dropped > 0:
        return f"{name}:\n{rendered}\n[{dropped} byte(s) truncated]"
    return f"{name}:\n{rendered}"

File: 03_use_cases/13_macos_cua/src/metacua/BashTool.swift
import Darwin
import Foundation

private let bashToolMaxOutputBytes = 128 * 1024

private final class ProcessOutputBuffer {
  private let lock = NSLock()
  private var data = Data()
  private var droppedByteCount = 0

  func append(_ chunk: Data) {
    guard !chunk.isEmpty else { return }

    lock.lock()
    defer { lock.unlock() }

    let remaining = bashToolMaxOutputBytes - data.count
    if remaining > 0 {
      data.append(chunk.prefix(remaining))
    }
    if chunk.count > remaining {
      droppedByteCount += chunk.count - max(remaining, 0)
    }
  }

  func snapshot() -> (text: String, droppedBytes: Int) {
    lock.lock()
    defer { lock.unlock() }

    let text = String(data: data, encoding: .utf8) ?? String(decoding: data, as: UTF8.self)
    return (text, droppedByteCount)
  }
}

func runBashTool(command: String, timeoutMS: Int) throws -> String {
  let trimmed = command.trimmingCharacters(in: .whitespacesAndNewlines)
  guard !trimmed.isEmpty else {
    throw CLIError("bash command must not be empty")
  }

  let process = Process()
  process.executableURL = URL(fileURLWithPath: "/bin/bash")
  process.arguments = ["-lc", command]
  process.currentDirectoryURL = URL(fileURLWithPath: FileManager.default.currentDirectoryPath)

  let stdoutPipe = Pipe()
  let stderrPipe = Pipe()
  let stdout = ProcessOutputBuffer()
  let stderr = ProcessOutputBuffer()
  let finished = DispatchSemaphore(value: 0)

  process.standardOutput = stdoutPipe
  process.standardError = stderrPipe
  stdoutPipe.fileHandleForReading.readabilityHandler = { handle in
    stdout.append(handle.availableData)
  }
  stderrPipe.fileHandleForReading.readabilityHandler = { handle in
    stderr.append(handle.availableData)
  }
  process.terminationHandler = { _ in
    finished.signal()
  }

  do {
    try process.run()
  } catch {
    stdoutPipe.fileHandleForReading.readabilityHandler = nil
    stderrPipe.fileHandleForReading.readabilityHandler = nil
    throw CLIError("failed to launch bash: \(error.localizedDescription)")
  }

  let waitResult = finished.wait(timeout: .now() + .milliseconds(timeoutMS))
  let timedOut = waitResult == .timedOut
  if timedOut {
    stdoutPipe.fileHandleForReading.readabilityHandler = nil
    stderrPipe.fileHandleForReading.readabilityHandler = nil
    if process.isRunning {
      process.terminate()
    }
    if finished.wait(timeout: .now() + .seconds(2)) == .timedOut, process.isRunning {
      kill(process.processIdentifier, SIGKILL)
      _ = finished.wait(timeout: .now() + .seconds(2))
    }
  }

  stdoutPipe.fileHandleForReading.readabilityHandler = nil
  stderrPipe.fileHandleForReading.readabilityHandler = nil
  stdout.append(stdoutPipe.fileHandleForReading.readDataToEndOfFile())
  stderr.append(stderrPipe.fileHandleForReading.readDataToEndOfFile())

  let result = formatBashToolResult(
    command: trimmed,
    timeoutMS: timeoutMS,
    timedOut: timedOut,
    exitCode: process.terminationStatus,
    stdout: stdout.snapshot(),
    stderr: stderr.snapshot()
  )

  if timedOut {
    throw CLIError(result)
  }
  if process.terminationStatus != 0 {
    throw CLIError(result)
  }
  return result
}

private func formatBashToolResult(
  command: String,
  timeoutMS: Int,
  timedOut: Bool,
  exitCode: Int32,
  stdout: (text: String, droppedBytes: Int),
  stderr: (text: String, droppedBytes: Int)
) -> String {
  var parts: [String] = []
  parts.append("command: \(command)")
  if timedOut {
    parts.append("status: timed out after \(timeoutMS)ms")
  } else {
    parts.append("status: exit \(exitCode)")
  }
  parts.append(formatStream(name: "stdout", stream: stdout))
  parts.append(formatStream(name: "stderr", stream: stderr))
  return parts.joined(separator: "\n")
}

private func formatStream(name: String, stream: (text: String, droppedBytes: Int)) -> String {
  let text = stream.text.trimmingCharacters(in: .newlines)
  let rendered = text.isEmpty ? "<empty>" : text
  if stream.droppedBytes > 0 {
    return "\(name):\n\(rendered)\n[\(stream.droppedBytes) byte(s) truncated]"
  }
  return "\(name):\n\(rendered)"
}
